armi/reactor/components/component.py

```diff
@@ -639,14 +636,49 @@ def getNumberDensity(self, nucName):
         number density : float
             number density in atoms/bn-cm.
         """
-        return self.p.numberDensities.get(nucName, 0.0)
+        i = np.where(self.p.nuclides == nucName.encode())[0]
+        if i.size > 0:
+            return self.p.numberDensities[i[0]]
+        else:
+            return 0.0
 
     def getNuclideNumberDensities(self, nucNames):
         """Return a list of number densities for the nuc names requested."""
-        return [self.p.numberDensities.get(nucName, 0.0) for nucName in nucNames]
+        if isinstance(nucNames, (list, tuple, np.ndarray)):
+            byteNucs = np.asanyarray(nucNames, dtype="S6")
+        else:
+            byteNucs = [nucName.encode() for nucName in nucNames]
+
+        nDens = np.zeros(len(byteNucs), dtype=np.float64)
+        if self.p.numberDensities is None:
+            return nDens
+
+        nuclideCopy = np.array(self.p.nuclides)
+        nDensCopy = np.array(self.p.numberDensities)
+
+        # trivial case where nucNames is the full set of nuclides in the same order
+        if np.array_equal(byteNucs, nuclideCopy):
+            return nDensCopy
+
+        if len(nDens) > len(nDensCopy) / 10:
+            # if there are a lot of indices to get densities for, use reverseIndex lookup
+            reverseIndex = {nuc: i for i, nuc in enumerate(self.p.nuclides)}
+            for i, nuc in enumerate(byteNucs):
+                j = reverseIndex.get(nuc, -1)
+                if j >= 0:
+                    nDens[i] = nDensCopy[j]
+        else:
+            # if it's just a small subset of nuclides, use np.where for direct index lookup
+            for i, nuc in enumerate(byteNucs):
+                j = np.where(nuclideCopy == nuc)[0]
+                if j.size > 0:
+                    nDens[i] = nDensCopy[j[0]]
+
+        return nDens
 
     def _getNdensHelper(self):
-        return dict(self.p.numberDensities)
+        nucs = self.getNuclides()
+        return dict(zip(nucs, self.p.numberDensities)) if len(nucs) > 0 else {}
```

Comment:
I can't help but wonder if some of your logic in this file would be simplified if you just let these methods return arrays instead of dictionaries now.

Thoughts?